consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+                        response.send(true).ok();
+                    }
+                    Err(e) => {
+                        error!("failed receiving block: {:?}", e);
+                        response.send(false).ok();
+                    }
+                },
+                ConsensusMessage::FetchMemoryPool(size, response) => {
+                    let out = self
+                        .memory_pool
+                        .get_candidates(size)
+                        .map(|x| x.into_iter().cloned().collect());
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreateTransaction(request, response) => {
+                    let out = self.create_transaction(*request);
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreatePartialTransaction(request, response) => {
+                    let out = self.create_partial_transaction(request);
+                    response.send(out).ok();
+                }
+            }
+        }
+    }
+
+    /// Check if the transaction is valid.
+    fn verify_transaction(&self, transaction: &SerialTransaction) -> Result<bool, ConsensusError> {
+        if !self
+            .public
+            .parameters
+            .authorized_inner_snark_ids
+            .contains(&to_bytes![transaction.inner_circuit_id]?)
+        {
+            return Ok(false);
+        }
+
+        Ok(InstantiatedDPC::verify(
+            &self.public.public_parameters,
+            &Tx::deserialize(transaction)?,
+            &self.ledger.deserialize::<Components>(),
+        )?)
+    }
+
+    /// Check if the transactions are valid.
+    fn verify_transactions(&self, transactions: &[SerialTransaction]) -> Result<bool, ConsensusError> {
+        for tx in transactions {
+            if !self
+                .public
+                .parameters
+                .authorized_inner_snark_ids
+                .iter()
+                .any(|x| &x[..] == &tx.inner_circuit_id[..])
+            {
+                return Ok(false);
+            }
+        }
+
+        let deserialized = transactions
+            .iter()
+            .map(|t| Tx::deserialize(t))
+            .collect::<IoResult<Vec<_>>>()?;
+
+        Ok(InstantiatedDPC::verify_transactions(
+            &self.public.public_parameters,
+            &deserialized[..],
+            &self.ledger.deserialize::<Components>(),
+        )?)
+    }
+
+    /// Check if the block is valid.
+    /// Verify transactions and transaction fees.
+    async fn verify_block(&self, block: &SerialBlock) -> Result<bool, ConsensusError> {
+        let transaction_ids: Vec<[u8; 32]> = block.transactions.iter().map(|x| x.id).collect();
+        let (merkle_root, pedersen_merkle_root, _) = txids_to_roots(&transaction_ids);
+
+        let canon = self.storage.canon().await?;
+        // Verify the block header
+        if block.header.previous_block_hash != canon.hash {
```

Comment:
This check should be moved to the very beginning, granting us with a cheaper potential early return.